### PR TITLE
test: Support for creating image refreshment pull requests

### PR DESCRIPTION
**Pull request by mvollmer**
This will be changed to create an issue instead of a pull request.

Depends on #3470.

For later:
- Commit and push the new image link

### CODE REVIEW

File: test/testinfra.py [hunk 1]
```diff
@@ -279,6 +292,17 @@ def statuses(self, revision):
                 count = len(statuses)
         return result
 
+    def pulls(self, state, limit=None):
+        result = [ ]
+        page = 1
+        count = 30
+        while count == 30 and (not limit or len(result) < limit):
+            pulls = self.get("pulls?state={0}&page={1}&per_page={2}".format(state, page, count))
```
Comment (stefwalter) [inline test/testinfra.py]: In theory pulls can be None.

File: test/testinfra.py [hunk 2]
```diff
@@ -264,6 +269,14 @@ def post(self, resource, data, accept=[]):
             raise Exception("GitHub API problem: {0}".format(response['reason'] or status))
         return json.loads(response['data'])
 
+    def delete(self, resource, data, accept=[]):
+        response = self.request("DELETE", resource, json.dumps(data), { "Content-Type": "application/json" })
+        status = response['status']
+        if (status < 200 or status >= 300) and status not in accept:
+            sys.stderr.write("{0}\n{1}\n".format(resource, response['data']))
+            raise Exception("GitHub API problem: {0}".format(response['reason'] or status))
```
Comment (stefwalter) [inline test/testinfra.py]: Should we handle 404 as a success? Due to races?

File: test/testinfra.py [hunk 3]
```diff
@@ -343,11 +367,18 @@ def scan(self, update, context, except_context=False):
         results = []
         master_contexts = []
         pull_contexts = []
+        verify_contexts = []
+        image_contexts = []
```
Comment (stefwalter) [inline test/testinfra.py]: I'm uncomortable merging this all into one monster function. There are several things that need to be worked out here:

 * Separating the code into more manageable chunks somehow.
 * Returning something other than a tuple, a GithubTask() object

File: test/testinfra.py [hunk 4]
```diff
@@ -343,11 +367,18 @@ def scan(self, update, context, except_context=False):
...
         for (context, what) in contexts.items():
```
Comment (stefwalter) [inline test/testinfra.py]: Also this function gets called every minute, are we sure that's how often we want to scan for creating new images? Or is there a better way to do that? Am I misunderstanding how this works?

File: test/testinfra.py (hunk 2 above)
Comment (mvollmer) [inline test/testinfra.py]: I'd say this is up to the caller, no?

File: test/testinfra.py (hunk 3 above)
Comment (mvollmer) [inline test/testinfra.py]: > I'm uncomortable merging this all into one monster function. 

I agree.

File: test/testinfra.py (hunk 4 above)
Comment (mvollmer) [inline test/testinfra.py]: I think you understand it correctly.

The extra work/traffic created by also fetching closed pull requests is not enourmous.  We don't get statuses or labels for all of them.

But yes, we can look for a better way.  Maybe a good old cron job or systemd timer is better?

Comment (mvollmer) [inline test/testinfra.py]: I'll extend `github-image-refresh` to do the scanning.

Comment (mvollmer) [inline test/testinfra.py]: Done

File: test/testinfra.py (hunk 2 above)
Comment (stefwalter) [inline test/testinfra.py]: I guess so. But there's a very common pattern of two callers doing a delete ... in which case both should succeed.

**Comment by mvollmer** [comment]
I'll start over...

**Pull request closed without merging**